File: packages/domain/src/claim-identity.ts
import { encodeAbiParameters, keccak256, stringToHex } from 'viem';
import type { Hex } from 'viem';
import { RequestValidationError } from './errors.js';

/**
 * Canonical type string for the claim usage derivation. It is the domain
 * separator: hashing it produces the first word, keeping this digest distinct
 * from request digests and claim commitments.
 */
export const CLAIM_USAGE_TYPE =
  'UltratokenizerClaimUsageV2(bytes32 sourceId,bytes32 claimId)';

/**
 * Authenticated fields from a signed source capsule that identify one claim.
 * `claimId` is the claim subject key: a stable, high-entropy, opaque source
 * claim identifier. None of these are holder wallets, salts or policy versions.
 */
export type ClaimIdentity = Readonly<{
  sourceId: Hex;
  claimId: Hex;
}>;

const identityFields = ['sourceId', 'claimId'] as const;

function identifier(value: unknown, field: string): Hex {
  if (
    typeof value !== 'string' ||
    !/^0x[0-9a-fA-F]{64}$/.test(value) ||
    /^0x0{64}$/.test(value)
  ) {
    throw new RequestValidationError(
      'invalid_identifier',
      field,
      `${field} must be a nonzero 32-byte hex value.`,
    );
  }
  return value.toLowerCase() as Hex;
}

/**
 * Stable, privacy-preserving claim usage identifier:
 *
 *   keccak256(abi.encode(typeHash, sourceId, claimId))
 *
 * It is a one-way hash of authenticated capsule identifiers, never a raw
 * document or account identifier, and it does not depend on any holder wallet,
 * issuer, salt or policy version. The caller must supply authenticated source fields;
 * this function validates their shape but not their authenticity.
 */
export function getClaimUsageId(input: unknown): Hex {
  if (
    input === null ||
    typeof input !== 'object' ||
    Array.isArray(input) ||
    (Object.getPrototypeOf(input) !== Object.prototype &&
      Object.getPrototypeOf(input) !== null)
  ) {
    throw new RequestValidationError(
      'invalid_shape',
      '$',
      'Claim identity must be a plain object.',
    );
  }
  const data = input as Record<string, unknown>;
  for (const key of Reflect.ownKeys(data)) {
    if (
      typeof key !== 'string' ||
      !identityFields.some((field) => field === key)
    ) {
      throw new RequestValidationError(
        'unknown_field',
        typeof key === 'string' ? key : '$',
        'Unsupported claim identity field.',
      );
    }
  }
  for (const field of identityFields) {
    if (!Object.hasOwn(data, field)) {
      throw new RequestValidationError(
        'missing_field',
        field,
        'Required claim identity field is missing.',
      );
    }
    if (!('value' in Object.getOwnPropertyDescriptor(data, field)!)) {
      throw new RequestValidationError(
        'invalid_shape',
        field,
        'Claim identity must contain data properties only.',
      );
    }
  }
  const sourceId = identifier(data.sourceId, 'sourceId');
  const claimId = identifier(data.claimId, 'claimId');
  return keccak256(
    encodeAbiParameters(
      [{ type: 'bytes32' }, { type: 'bytes32' }, { type: 'bytes32' }],
      [keccak256(stringToHex(CLAIM_USAGE_TYPE)), sourceId, claimId],
    ),
  );
}

File: proofs/claim-evidence/src/claim_identity.rs
//! Canonical claim usage identity, identical to packages/domain/claim-identity.ts.

use crate::{hash, request::hash_words, request::Word};

/// EIP-712-style type string; its hash is the domain separator for this
/// derivation, keeping it distinct from request digests and claim commitments.
pub const CLAIM_USAGE_TYPE: &[u8] = b"UltratokenizerClaimUsageV2(bytes32 sourceId,bytes32 claimId)";

/// Stable, privacy-preserving claim usage identifier:
///
///   keccak256(abi.encode(typeHash, sourceId, claimId))
///
/// It is a one-way hash of authenticated capsule identifiers, never a raw
/// document or account identifier, and it does not depend on the mint issuer,
/// holder wallet, signing key, salt or policy version. Zero-word rejection is
/// the caller's boundary, not part of this hash.
pub fn claim_usage_id(source_id: Word, claim_id: Word) -> Word {
    // Zero-word rejection is enforced by callers (capsule parse, CLI decode);
    // assert the invariant here so a direct misuse fails loudly in tests.
    debug_assert!(
        source_id != [0; 32] && claim_id != [0; 32],
        "claim identity words must be nonzero"
    );
    hash_words(&[hash(CLAIM_USAGE_TYPE), source_id, claim_id])
}
